keanu-project/src/main/java/io/improbable/keanu/model/LogisticRegression.java

```diff
@@ -0,0 +1,111 @@
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+
+public class LogisticRegression implements LinearModel {
+
+    private static final double DEFAULT_MU = 0.0;
+    private static final double DEFAULT_SIGMA = 2.0;
+    private static final double DEFAULT_REGULARIZATION = 1;
+    private static final VertexLabel WEIGHTS_LABEL = new VertexLabel("weights");
+    private static final VertexLabel INTERCEPT_LABEL = new VertexLabel("intercept");
+    private static final VertexLabel PROBABILITIES_LABEL = new VertexLabel("probabilities");
+    private static final VertexLabel OUTCOMES_LABEL = new VertexLabel("outcomes");
+
+    private DoubleTensor x;
+    private DoubleTensor y;
+    private BayesianNetwork net;
+    private double priorMu;
+    private double priorSigma;
+    private boolean isFit = false;
+    private double regularization;
+
+
+    public LogisticRegression(DoubleTensor x, DoubleTensor y, double regularization) {
+        this.x = x;
+        this.y = y;
+        this.priorMu = DEFAULT_MU;
+        this.priorSigma = DEFAULT_SIGMA;
+        this.regularization = regularization;
+        construct();
+    }
+
+    public LogisticRegression(DoubleTensor x, DoubleTensor y) {
+        this(x, y, DEFAULT_REGULARIZATION);
+    }
+
+    @Override
+    public BayesianNetwork construct() {
+        DoubleVertex probabilities = computeProbabilities(x);
+        probabilities.setLabel(PROBABILITIES_LABEL);
+        BernoulliVertex outcomes = new BernoulliVertex(probabilities);
+        outcomes.setLabel(OUTCOMES_LABEL);
+        net = new BayesianNetwork(outcomes.getConnectedGraph());
+        return net;
+    }
+
+    private DoubleVertex computeProbabilities(DoubleTensor x) {
+        int numFeatures = x.getShape()[0];
+        double[] sigma = new double[numFeatures];
+        for (int i = 0; i < numFeatures; i++) {
+            DoubleTensor xColumn = x.slice(0, i);
+            sigma[i] = priorSigma / xColumn.abs().average() / regularization;
+        }
+        DoubleVertex intercept = new GaussianVertex(priorMu, priorSigma);
+        intercept.setLabel(INTERCEPT_LABEL);
+        DoubleVertex weights = new GaussianVertex(priorMu, ConstantVertex.of(sigma));
+        weights.setLabel(WEIGHTS_LABEL);
+
+        ConstantDoubleVertex xVertex = ConstantVertex.of(x);
+        return weights.getValue().isScalar() ?
+            weights.times(xVertex).plus(intercept).sigmoid() :
+            weights.matrixMultiply(xVertex).plus(intercept).sigmoid();
+    }
+
+    @Override
+    public LogisticRegression fit() {
+        BooleanTensor observedOutcomes = y.greaterThan(0.5);
+        net.getVertexByLabel(OUTCOMES_LABEL).observe(observedOutcomes);
+        GradientOptimizer optimizer = GradientOptimizer.of(net);
+        optimizer.maxAPosteriori();
```

Comment:
why do we do `maxAPosteriori` for `LogisticRegression` but `maxLikelihood` for `LinearRegression`?

Reply:
My mistake - changed to MAP on both.